### PR TITLE
changing storeKeyconverter logic to count based logic

### PR DESCRIPTION
**Pull request by manbearpig1996**
Added tryDropCache for keys in StoreKeyConverter to try and drop some keys.
Adding an AbstractStoreKeyConverter and implementing in current test classes of StoreKeyConverter
to have same expected logic everywhere.
AbstractStoreKeyConverter is using a counting cache logic for keys.
Whenever a key is added its count will keep increasing, when tryDropCache is called for keys, count will decrease.
Keys will be dropped, when count becomes 0.
Rest of the behviour is same as expected

### CODE REVIEW

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

File: ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    StoreKey storeKey1 = mock(StoreKey.class);
+    assertNotSame("storeKeys should not be equal", storeKey0, storeKey1);
+    List<StoreKey> list = Arrays.asList(storeKey0, storeKey1);
+
+    try {
+      storeKeyConverter.getConverted(storeKey0);
+      fail("Get succeeded for not converted keys");
+    } catch (Exception e) {
+      //expected
+    }
```
Comment (mudit-saxena) [inline ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java:71]: Can you out this as a separate test method if this is expected to throw exception? Also, preferably use `assertThrows` or `@Test(expected = ExceptionClass.class)`

File: ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+    Map<StoreKey, StoreKey> convertedMap = storeKeyConverter.convert(list);
+    assertEquals("Size of keys in input and output is different", list.size(), convertedMap.size());
+
+    try {
+      StoreKey storeKey0Res = storeKeyConverter.getConverted(storeKey0);
+      assertNotNull("Conversion of storeKey0 should not be null", storeKey0Res);
+    } catch (Exception e) {
+      fail("Exception thrown while storeKey0 is present");
```
Comment (mudit-saxena) [inline ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java:80]: If you expect test failure when an exception is thrown you can avoid catching it

File: ambry-api/src/main/java/com/github/ambry/store/AbstractStoreKeyConverter.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  protected abstract Map<StoreKey, StoreKey> convertKeys(Collection<? extends StoreKey> input) throws Exception;
+
+  /**
+   * This methods response will be returned from {@link #getConverted(StoreKey)} method.
+   * @param storeKey  StoreKey that was queried
+   * @param isKeyPresent whether key is present in cache
+   * @param cachedMapping convertedMapping that is stored in cache. Will be null if if key is not present in cache.
+   * @return
+   */
+  protected abstract StoreKey getConvertedKey(StoreKey storeKey, Boolean isKeyPresent, StoreKey cachedMapping);
```
[the remark this answers is not in the document]
Comment (manbearpig1996) [reply ambry-api/src/main/java/com/github/ambry/store/AbstractStoreKeyConverter.java]: Changed it.

File: ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java (hunk 1 above)
Comment (manbearpig1996) [reply ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java:71]: Need to have a test where first we are throwing exception then we are not and then we are again throwing and so on. So need to write code in a single place this way.
assertThrows is not present in junit4 that we are using in ambry-oss.

File: ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java (hunk 2 above)
Comment (manbearpig1996) [reply ambry-commons/src/test/java/com/github/ambry/store/AbstractStoreKeyConverterTest.java:80]: Need to have a test where first we are throwing exception then we are not and then we are again throwing and so on. So need to write code in a single place this way.

**Review by mudit-saxena — APPROVED** [review]
GTG

**Pull request merged**